enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -3078,19 +3255,19 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
   }
 
   if (hasMetadata(key.todiff, "enzyme_gradient")) {
-
-    DIFFE_TYPE subretType = key.todiff->getReturnType()->isFPOrFPVectorTy()
-                                ? DIFFE_TYPE::OUT_DIFF
-                                : DIFFE_TYPE::DUP_ARG;
+    std::set<llvm::Type *> seen;
+    DIFFE_TYPE subretType = whatType(key.todiff->getReturnType(),
+                                     DerivativeMode::ReverseModeGradient,
+                                     /*intAreConstant*/ false, seen);
     if (key.todiff->getReturnType()->isVoidTy() ||
         key.todiff->getReturnType()->isEmptyTy())
       subretType = DIFFE_TYPE::CONSTANT;
     assert(subretType == key.retType);
 
-    auto res = getDefaultFunctionTypeForGradient(key.todiff->getFunctionType(),
-                                                 /*retType*/ key.retType);
-
     if (key.mode == DerivativeMode::ReverseModeCombined) {
+      auto res = getDefaultFunctionTypeForGradient(
```

Comment:
Please use a structured binding to unwrap the pair elements: 
`auto&& [nameForFirstPairElement, nameForSecondPairElement] = ...`